labs/MLH/stack.rst

```diff
@@ -0,0 +1,258 @@
+.. meta::
+  :author: AdaCore
+
+:prev_state: False
+:next_state: False
+
+:code-config:`accumulate_code=False;reset_accumulator=True`
+
+Let's Build a Stack
+=====================
+
+In this lab we will build a stack data structure and use the SPARK provers to find the errors in the below implementation.
+
+Background
+----------
+
+**So, what is a stack?**
+
+A stack is like a pile of dishes...
+
+.. image:: pile_of_dishes.png
+   :align: center
+   :scale: 55 %
+
+#. The pile starts out empty.
+#. You add ( :ada:`push` ) a new plate ( :ada:`data` ) to the stack by placing it on the top of the pile.
+#. To get plates ( :ada:`data` ) out, you take the one off the top of the pile ( :ada:`pop` ).
+#. Out stack has a maximum height ( :ada:`size` ) of 9 dishes
+
+
+**Pushing items onto the stack**
+
+Here's what should happen if we pushed the string :ada:`MLH` onto the stack.
+
+.. container:: img_row
+
+   .. image:: push_1.png
+      :scale: 50 %
+
+   .. image:: push_2.png
+      :scale: 50 %
+
+   .. image:: push_3.png
+      :scale: 50 %
+
+   .. image:: push_4.png
+      :scale: 50 %
+
+   .. image:: push_5.png
+      :scale: 50 %
+
+The list starts out empty. Each time we push a character onto the stack, :ada:`Last` increments by :ada:`1`.
+
+**Popping items from the stack**
+
+Here's what should happen if we popped :ada:`2` characters off our stack & then clear it.
+
+.. container:: img_row
+
+   .. image:: pop_1.png
+      :scale: 50 %
+
+   .. image:: pop_2.png
+      :scale: 50 %
+
+   .. image:: pop_3.png
+      :scale: 50 %
+
+   .. image:: pop_4.png
+      :scale: 50 %
+
+Note that :ada:`pop` and :ada:`clear` don't unset the :ada:`Storage` array's elements, they just change the value of :ada:`Last`.
+
+Input Format
+------------
+
+N inputs will be read from stdin/console as inputs, C to the stack.
+
+Constraints
+-----------
+
+1 <= N <= 1000
+C is any character. Characters d and p will be special characters corresponding to the below commands:
+
+p => Pops a character off the stack
+
+d => Prints the current characters in the stack
+
+Output Format
+-------------
+
+If the stack currently has the characters "M", "L", and "H" then the program should print the stack like this:
+
+[M, L, H]
+
+Sample Input
+------------
+
+M L H d p d p d p d
+
+Sample Output
+-------------
+
+[M, L, H]
+[M, L]
+[M]
+[]
+
+--------------
+
+
+.. code:: ada lab=MLH_Stack
+
+   --  START LAB IO BLOCK
+   in 0: M L H d p d p d p d
+   out 0: [M, L, H] [M, L] [M] []
+   in 1: a b c d e d p
+   out 1: [a, b, c] [a, b, c, e]
+   in 2: p a p d
+   out 2: Nothing to Pop, Stack is empty! []
+   --  END LAB IO BLOCK
+
+   package Stack with SPARK_Mode => On is
```

Comment:
Editorial remark: we're using 4 spaces everywhere in the courses as the indentation of the 'code' blocks. Of course, the indentation of the actual Ada code is still the familiar one with 3 spaces.

Reply:
Editorial remark (2): there are trailing whitespaces in some files. We should try to avoid that --- most editors support removing them before saving files.